src/main/java/org/candlepin/subscriptions/product/StubProductApi.java

```diff
@@ -0,0 +1,79 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.product;
+
+import org.candlepin.subscriptions.product.api.model.EngineeringProductMap;
+import org.candlepin.subscriptions.product.api.model.RESTProductTree;
+import org.candlepin.subscriptions.product.api.model.SkuEngProduct;
+import org.candlepin.subscriptions.product.api.resources.ProductApi;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Stub version of {@link ProductApi} for {@link ProductApiFactory} for local testing.
+ */
+public class StubProductApi extends ProductApi {
+
+    // Get the same type of ObjectMapper as used by ProductApi.
+    private static final ObjectMapper MAPPER = (new JSON()).getContext(ObjectMapper.class);
+
+    @Override
+    public EngineeringProductMap getEngineeringProductsForSkus(String sku) throws ApiException {
+        List<String> skuList = List.of(sku.split(","));
+        List<SkuEngProduct> sepList = new ArrayList<>(skuList.size());
+
+        for (String skuItem : skuList) {
+            String resName = String.format("engprods-%s.json", skuItem);
+            EngineeringProductMap prodMap = readJsonResource(resName, EngineeringProductMap.class);
+            if (prodMap != null && prodMap.getEntries() != null && !prodMap.getEntries().isEmpty()) {
+                SkuEngProduct sep = prodMap.getEntries().get(0);
+                sepList.add(sep);
+            }
+        }
+
+        EngineeringProductMap epm = new EngineeringProductMap();
+        epm.setEntries(sepList);
+
+        return epm;
+    }
+
+    @Override
+    public RESTProductTree getProductTree(String sku, Boolean attributes) throws ApiException {
+        String resName = String.format("tree-%s_attrs-%b.json", sku, attributes);
+        return readJsonResource(resName, RESTProductTree.class);
+    }
+
+    private <T> T readJsonResource(String resName, Class<T> clazz) throws ApiException {
+        try (InputStream res = StubProductApi.class.getResourceAsStream(resName)) {
+            return MAPPER.readValue(res, clazz);
+        }
+        catch (IOException e) {
+            throw new ApiException(e);
+        }
+
+    }
```

Comment:
Could you please:
- [ ] throw mocked examples into the codebase and document it (comment here is fine)?
- [ ] document (internally?) how to produce these files from real data; add a pointer in comments of where to look for this documentation.
- [ ] handle gracefully missing files (return empty/null object?)

Reply:
Can do! I originally had two products pulled from QA-env's product service, so I can put it back. As for the final bullet, I should make this stub product api behave like the real product service anyway, which will succeed and return a null object anyway, so I'll do that :)

Reply:
I've completed the changes:

- "Mocked" examples (these files are actually from real product data from QA, but are stripped to only the data we use from the service). These files are located in src/main/resources/org/candlepin/subscriptions/product/ so if you'd rather them be somewhere other than main, let me know.
- I've provided a document on how to produce these files, accessible internally at https://source.redhat.com/groups/public/teamnado/blog/grabbing_real_sku_tree_and_engprod_oid_data_for_swatch_stubproductapi
- Missing files will return a null object. This better aligns to what the real product service does for these calls anyway, so that's nice :)